### PR TITLE
Refactor QPDFTokenizer

### PR DESCRIPTION
**Pull request by m-holger**
Reduce copying of strings by:

- use of std::string_view
- retrieving results on the fly as needed instead of copying them into Token objects
- not constructing both QPDFTokenizer::val and raw_val  when they are identical

### CODE REVIEW

File: include/qpdf/QUtil.hh [hunk 1]
```diff
@@ -535,6 +536,12 @@ namespace QUtil
     // purposes.
     QPDF_DLL
     size_t get_max_memory_usage();
+
+    // Temporary methods to be removed in qpdf 12.
```
Comment (jberkenbilt) [inline include/qpdf/QUtil.hh:540]: Why are they being removed in qpdf 12? What changes do you anticipate in qpdf 12 that will make these no longer needed?

**Review by jberkenbilt — CHANGES_REQUESTED** [review]

**Comment by jberkenbilt** [comment]
It will take me a little more time to study this before I can be sure whether I like it, but it seems good at first glance. I've never really liked the part of QPDFTokenizer that you're working with -- that interface has been hard to use, but my earlier attempt to kill it (years ago when I tried to write the tokenizer using flex) made the performance worse. I didn't have std::string_view at the time though. Anyway, if you make claims about new public API being temporary, I'll need more detail about where you're planning on going with it.

File: include/qpdf/QUtil.hh (hunk 1 above)
Comment (m-holger) [reply include/qpdf/QUtil.hh:540]: There are existing versions of these methods that take std::string const& parameters which can be transparently changed to take string_view parameters (which cover string_view, std::strings and C strings) except this cannot happen until  qpdf 12 as it is ABI breaking.

Comment (m-holger) [reply include/qpdf/QUtil.hh:540]: It is not possible to overload the existing method with versions that take string_views as the compiler is unable to work out which version to use.

Comment (jberkenbilt) [reply include/qpdf/QUtil.hh:540]: The way I would do this (and request for you to do it) is, for each function, change to something like this:
```c++
    // ABI: change std::string const& to std::string_view (API-compatible) and remove the temporary view version.
    QPDF_DLL
    bool utf8_to_pdf_doc(
        std::string const& utf8, std::string& pdfdoc, char unknown_char = '?');
    // temporary until next ABI change
    bool utf8_view_to_pdf_doc(
        std::string_view utf8, std::string& pdfdoc, char unknown_char = '?');
```
See `ABI Changes` in TODO. Following this convention makes it very easy to find and address these the next time the ABI window opens up.

**Comment by m-holger** [comment]
I have taken out commits depending on string_view.

I have also made the changes private (until qpdf 12).

File: include/qpdf/QPDFTokenizer.hh [hunk 2]
```diff
@@ -219,7 +220,17 @@ class QPDFTokenizer
     bool nextToken(
         InputSource& input, std::string const& context, size_t max_len = 0);
 
-  private:
+    // The following methods are only valid after nextToken has been called
+    // and until another QPDFTokenizer method is called. They allow the results
+    // of calling nextToken to be accessed without creating a Token, thus
+    // avoiding copying information that may not be needed.
+    // NB Any string_views returned by these methods will be invalid after
```
[the remark this answers is not in the document]
Comment (jberkenbilt) [reply include/qpdf/QPDFTokenizer.hh:227]: Maybe we should never return std::string_view at all.

**Review by jberkenbilt — CHANGES_REQUESTED** [review]
Looks good except for one question and some stray comments. Also please note about my comments about string_view for future reference. Let me know your thoughts.

File: include/qpdf/QPDFTokenizer.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (jberkenbilt) [reply include/qpdf/QPDFTokenizer.hh:227]: I added some notes about this to #785.

[the remark this answers is not in the document]
Comment (jberkenbilt) [reply include/qpdf/QPDFTokenizer.hh:227]: I think this might be a case where returning std::string_view would be okay since this is a very specialized interface and performance critical, but if we do that, we should do it now and not start with `std::string const&` and change it later.

[the remark this answers is not in the document]
Comment (jberkenbilt) [reply include/qpdf/QPDFTokenizer.hh:227]: And your NB comment provides the needed warning.

File: libqpdf/QPDFTokenizer.cc [hunk 3]
```diff
@@ -342,22 +333,20 @@ QPDFTokenizer::inTop(char ch)
...
         return;
 
     case '.':
         this->state = st_decimal;
-        this->val += ch;
         return;
 
     default:
         this->state = st_literal;
+        // We need to build val to test it for "null", "true" of "false"
```
[the remark this answers is not in the document]
Comment (m-holger) [reply libqpdf/QPDFTokenizer.cc:349]: Well spotted.

File: include/qpdf/QPDFTokenizer.hh [hunk 4]
```diff
@@ -30,6 +30,7 @@
 #include <memory>
 #include <stdio.h>
 #include <string>
+#include <string_view>
```
[the remark this answers is not in the document]
Comment (m-holger) [reply include/qpdf/QPDFTokenizer.hh:33]: Removed

File: include/qpdf/QPDFTokenizer.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (m-holger) [reply include/qpdf/QPDFTokenizer.hh]: > Are any string_views returned?

Not any longer. Comment removed.

> Maybe we should never return std::string_view at all.

Agreed for the public API.

For the private API it there may be a case, but it should limited to
- where it is performance critical, and
- where the string_view is immediately consumed.

I suspect that there probably is no need to return string_views in qpdf across class boundaries.

Both for the public and private API there are probably cases where we should return a string&& instead of a string const&.

File: include/qpdf/QPDFTokenizer.hh (hunk 4 above)
[the remark this answers is not in the document]
Comment (jberkenbilt) [reply include/qpdf/QPDFTokenizer.hh:33]: It still shows up in the diff. Is it a github problem or is not actually removed?

**Comment by jberkenbilt** [comment]
Looks like the string_view include is still there. Will hold until that's removed unless you feel like these interfaces really deserve to return `std::string_view` since they are performance-critical and new. I'm open to that with the disclaimer you had previously put there. Or if you're good with just having it be `std::string const&`, then just remove the include, and this is good to go.

File: include/qpdf/QPDFTokenizer.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (jberkenbilt) [reply include/qpdf/QPDFTokenizer.hh]: I never considered returning a `string&&`. I guess that lets you return a temporary string without copying it? And I suppose auto would still be string. I guess since this is a public API, we can leave it as `std::string const&` for now.

**Review by jberkenbilt — APPROVED** [review]
Looks good -- will merge when CI passes unless you have anything else to discuss about it.
